In [1]:
import pandas as pd
import numpy as np
import os,glob
from matplotlib import pyplot as plt

In [2]:
import time

In [3]:
from ldacgs import *

In [4]:
path = 'other_files/'
f = 'merged_summary_topics.json'

In [5]:
new = pd.read_json(path + f)
# my data is in data2 not data
new['File'] = [x[5:] for x in new.File]
new['File_num'] = [int(x[:-4]) for x in new.File]
new['tag_count'] = [len(x) for x in new.topic_lists]

In [6]:
keep = ['Title','Speaker','File','Year','Month_letter','topic_lists','tag_count','File_num']
new.loc[new.anger == 1][keep]

Title                Speaker  \
7                                     Be Slow to Anger  ElRay L. Christiansen   
1955  Living Worthy of the Girl You Will Someday Marry     Gordon B. Hinckley   
1965                                  Agency and Anger        Lynn G. Robbins   
2019           Strengthening Families: Our Sacred Duty        Robert D. Hales   
2056                                        Our Legacy      Stephen B. Oveson   
2091                       Keep an Eternal Perspective          Jay E. Jensen   
2210                     Reaching Down to Lift Another     Gordon B. Hinckley   
2301        Forgiveness Will Change Bitterness to Love      David E. Sorensen   
2644                                     Slow to Anger     Gordon B. Hinckley   
2791                 School Thy Feelings, O My Brother       Thomas S. Monson   

          File  Year Month_letter  \
7     2009.txt  1971            A   
1955  6067.txt  1998            A   
1965  6077.txt  1998            A   
2019  6210.txt  1999            A   
2056  6285.txt  1999            O   
2091  6358.txt  2000            A   
2210  6594.txt  2001            O   
2301  6802.txt  2003            A   
2644  7497.txt  2007            O   
2791  7681.txt  2009            O   

                                            topic_lists  tag_count  File_num  
7                                  [anger, forgiveness]          2      2009  
1955  [missionary work, anger, education, loyalty, p...         16      6067  
1965  [Jesus Christ, abuse, anger, accountability, a...          7      6077  
2019  [councils, family history, prophets, anger, fa...         13      6210  
2056  [Jesus Christ, faith, Sunday School, tithing, ...         12      6285  
2091  [missionary work, faith, music, perspective, s...         15      6358  
2210  [anger, commandments, education, priesthood, w...          7      6594  
2301      [anger, forgiveness, love, criticism, family]          5      6802  
2644                         [divorce, anger, marriage]          3      7497  
2791             [anger, priesthood, self-control, sin]          4      7681

In [7]:
file_num = 7681
x = new.loc[new.File_num == file_num][['Title','Speaker','File','Year','Month','topic_lists','File_num','tag_count']]
print(x.topic_lists.values[0])
x

['anger', 'priesthood', 'self-control', 'sin']


Title           Speaker      File  Year  \
2791  School Thy Feelings, O My Brother  Thomas S. Monson  7681.txt  2009   

      Month                             topic_lists  File_num  tag_count  
2791     10  [anger, priesthood, self-control, sin]      7681          4

In [8]:
new.loc[new.File_num == file_num].File.values[0]

'7681.txt'

# Iterate through the talks
-

In [14]:
start = time.time()
time.sleep(5)
print(time.time() - start)

5.003112077713013


In [20]:
new.columns[:10]

Index(['Year', 'Month_letter', 'Speaker', 'Title', 'File', 'Kicker', 'Month',
       'Unnamed: 0', 'topic_lists', 'Aaronic Priesthood'],
      dtype='object')

In [31]:
import sys
import traceback



In [39]:
counter = 0
files = new.loc[new.fasting == 1].File_num.values
repeat_lda_times = 5
total_times = len(files) * repeat_lda_times
print(f'this is going to run {total_times} times')

this is going to run 150 times


# BECAUSE I'M GOING BY TOPIC - I COULD ACCIDENTALLY BE REPEATING TALKS IN THE LDA that have tags in both

In [40]:


# results = dict()
# beginning = time.time()
# errors = 0
# for file_num in files:
#     for i in range(repeat_lda_times):
#         try:
#             x = new.loc[new.File_num == file_num]
#             start = time.time()
#             #gs for gibbs sampler
#             gs = LDACGS(n_topics=10)
#             # gs.buildCorpus(filename='reagan.txt')
#             # gs.initialize()
#             # gs._sweep()

#             gs.sample(filename='data2/' + new.loc[new.File_num == file_num].File.values[0]
#                     , stopwords_file='../3winter21/Gibbs_LDA/' + 'stopwords.txt')

#             results[counter] = {'File_num':file_num
#                  ,'Title':x.Title
#                  ,'Speaker':x.Speaker
#                  ,'Year':x.Year
#                  ,'Month':x.Month
#                   #iter is number of times LDA was run on this talk 
#                  ,'iter':i
#                  ,'time (sec)':time.time() - start
#                  ,'log_probs':gs.logprobs
#                  ,'fifty_words':np.array(gs.topterms(n_terms=5))
#                  }
#             counter += 1
#         except:
#             errors += 1
#             message = str({'File_num':file_num
#              ,'Title':x.Title
#              ,'Speaker':x.Speaker
#              ,'Year':x.Year
#              ,'Month':x.Month })
#             traceback.print_exc()
#             print(message, 'failed')
#     if (counter % 5) == 0:
#         print(f'done with counter = {counter},or {np.round(counter / total_times * 100,1)}% done')
#         print(f'its been running for {np.round((time.time() - beginning) / 60,2)} minutes')
# print(f'there were {errors} errors')
        

done with counter = 5,or 3.3% done
its been running for 0.63 minutes
done with counter = 10,or 6.7% done
its been running for 1.03 minutes
done with counter = 15,or 10.0% done
its been running for 1.45 minutes
done with counter = 20,or 13.3% done
its been running for 1.56 minutes
done with counter = 25,or 16.7% done
its been running for 2.27 minutes
done with counter = 30,or 20.0% done
its been running for 2.45 minutes
done with counter = 35,or 23.3% done
its been running for 2.88 minutes
done with counter = 40,or 26.7% done
its been running for 3.37 minutes
done with counter = 45,or 30.0% done
its been running for 4.22 minutes
done with counter = 50,or 33.3% done
its been running for 4.81 minutes
done with counter = 55,or 36.7% done
its been running for 5.44 minutes
done with counter = 60,or 40.0% done
its been running for 6.21 minutes
done with counter = 65,or 43.3% done
its been running for 6.86 minutes
done with counter = 70,or 46.7% done
its been running for 7.51 minutes
done with

In [ ]:
chosen_topic = 'fasting'
repeat_lda_times = 5
params = [(0.001,1)
    ,(0.005,1)
    ,(0.001,0.17)]
attempt = 0
output_filename = f'{chosen_topic}_{repeat_lda_times}_lda_per_talk_attempt_{attempt}.json'

In [ ]:
results = dict()
beginning = time.time()
errors = 0

counter = 0
files = new.loc[new[chosen_topic] == 1].File_num.values
# new_set = set(files)
# I don't need to worry about repreats right now because I'm exploring with alpha beta params
# files = new_set - already_done_file_nums
# already_done_file_nums = already_done_file_nums.union(new_set)

total_times = len(files) * repeat_lda_times
print(f'this is going to run {total_times} times')
for file_num in files:
    for i in range(repeat_lda_times):
        # avoid out of error with accessing betas
        j = (i % len(params))
        a,b = params[j]
        raise
        try:
            x = new.loc[new.File_num == file_num]
            start = time.time()
            #gs for gibbs sampler
            gs = LDACGS(n_topics=10,alpha=a,beta=b)
            # gs.buildCorpus(filename='reagan.txt')
            # gs.initialize()
            # gs._sweep()

            gs.sample(filename='data2/' + new.loc[new.File_num == file_num].File.values[0]
                    , stopwords_file='../3winter21/Gibbs_LDA/' + 'stopwords.txt')

            results[counter] = {'File_num':file_num
                 ,'Title':x.Title
                 ,'Speaker':x.Speaker
                 ,'Year':x.Year
                 ,'Month':x.Month
                 ,'alpha':a
                 ,'beta':b
                 ,'topic_lists':x.topic_lists
                  #iter is number of times LDA was run on this talk
                 ,'iter':i
                 ,'time (sec)':time.time() - start
                 ,'log_probs':gs.logprobs
                 ,'fifty_words':np.array(gs.topterms(n_terms=5))
                 }
            counter += 1
        except:
            errors += 1
            message = str({'File_num':file_num
             ,'Title':x.Title
             ,'Speaker':x.Speaker
             ,'Year':x.Year
             ,'Month':x.Month })
            traceback.print_exc()
            print(message, 'failed')
    if (counter % 5) == 0:
        print(f'done with counter = {counter},or {np.round(counter / total_times * 100,1)}% done')
        print(f'its been running for {np.round((time.time() - beginning) / 60,2)} minutes')
print(f'there were {errors} errors')

pd.DataFrame(results).T.to_json(output_filename)

# How long did it take 
- pride, 5 per talk lda runs, 135 runs total: 14 minutes
- fasting,  5 per talk lda runs, 150 runs total: 16 minutes (0 errors)

# add in alpha, beta into results, and should I change it? 

In [41]:
attempt = 0
pd.DataFrame(results).T.to_json(f'fasting_5_lda_per_talk_attempt_{attempt}.json')

In [42]:
glob.glob('*.json')

['pride_5_lda_per_talk_attempt_0.json',
 'anger_5_lda_per_talk_attempt_0.json',
 'baptism_5_lda_per_talk_attempt_0.json',
 'fasting_5_lda_per_talk_attempt_0.json']

In [27]:
# then for each time that lda was run for one talk, combine all the topterms in to one feature space, 
r = pd.DataFrame(results).T

# look for talks that have been repeated

In [43]:
jsons = dict()
for f in glob.glob('*.json'):
    jsons[f] = pd.read_json(f)

In [44]:
#t for tempt
t = jsons['pride_5_lda_per_talk_attempt_0.json'].copy()
t.columns

Index(['File_num', 'Title', 'Speaker', 'Year', 'Month', 'iter', 'time (sec)',
       'log_probs', 'fifty_words'],
      dtype='object')

In [50]:
file_nums = np.array([])
for f in glob.glob('*.json'):
    t = jsons[f].copy()
    file_nums = np.append(file_nums,list(set(t.File_num.values)))

In [51]:
unique_elements, counts_elements = np.unique(np.array(file_nums), return_counts=True)

In [58]:
# this shows that each talk that I've done this far hasn't been repeated
already_done_file_nums = set(unique_elements)

In [62]:
print(set(unique_elements))

{6144.0, 7681.0, 6658.0, 5379.0, 8072.0, 5769.0, 5002.0, 5003.0, 6285.0, 8078.0, 7055.0, 3088.0, 8079.0, 6802.0, 6292.0, 6296.0, 5017.0, 4250.0, 8216.0, 4259.0, 6520.0, 5925.0, 6055.0, 6059.0, 5680.0, 6067.0, 5303.0, 3512.0, 6072.0, 6583.0, 7610.0, 6077.0, 3006.0, 8127.0, 6210.0, 4931.0, 6594.0, 7877.0, 3529.0, 5450.0, 5835.0, 7497.0, 7754.0, 5454.0, 6606.0, 5840.0, 6609.0, 8268.0, 8270.0, 4948.0, 7764.0, 6358.0, 7638.0, 2009.0, 5850.0, 8155.0, 4700.0, 2528.0, 5472.0, 8033.0, 8036.0, 8165.0, 2022.0, 4198.0, 6502.0, 8169.0, 6508.0, 5230.0, 6511.0, 8049.0, 8053.0, 8054.0, 5240.0, 6525.0, 7039.0}


In [66]:
new_set = {7764.0, 6358.0, 7638.0,-1,-2,-3}
#use set difference


{-3, -2, -1}

In [63]:
x.columns[:10]

Index(['Year', 'Month_letter', 'Speaker', 'Title', 'File', 'Kicker', 'Month',
       'Unnamed: 0', 'topic_lists', 'Aaronic Priesthood'],
      dtype='object')